launcher/InstanceCopyTask.cpp

```diff
@@ -5,18 +5,66 @@
 #include "pathmatcher/RegexpMatcher.h"
 #include <QtConcurrentRun>
 
-InstanceCopyTask::InstanceCopyTask(InstancePtr origInstance, bool copySaves, bool keepPlaytime)
+InstanceCopyTask::InstanceCopyTask(InstancePtr origInstance, const InstanceCopyPrefs& prefs)
 {
     m_origInstance = origInstance;
-    m_keepPlaytime = keepPlaytime;
+    m_keepPlaytime = prefs.keepPlaytime;
+    QString filter;
 
-    if(!copySaves)
+    if(!prefs.copySaves)
     {
-        // FIXME: get this from the original instance type...
-        auto matcherReal = new RegexpMatcher("[.]?minecraft/saves");
-        matcherReal->caseSensitive(false);
-        m_matcher.reset(matcherReal);
+        appendToFilter(filter, "saves");
     }
+
+    if(!prefs.copyGameOptions) {
+        appendToFilter(filter, "options.txt");
+    }
+
+    if(!prefs.copyResourcePacks)
+    {
+        appendToFilter(filter, "resourcepacks");
+        appendToFilter(filter, "texturepacks");
+    }
+
+    if(!prefs.copyShaderPacks)
+    {
+        appendToFilter(filter, "shaderpacks");
+    }
+
+    if(!prefs.copyServers)
+    {
+        appendToFilter(filter, "servers.dat");
+        appendToFilter(filter, "servers.dat_old");
+        appendToFilter(filter, "server-resource-packs");
+    }
+
+    if(!prefs.copyMods)
+    {
+        appendToFilter(filter, "coremods");
+        appendToFilter(filter, "mods");
+        appendToFilter(filter, "config");
+    }
```

Comment:
since you made a struct with the preferences, wouldn't it be better to have a function like 'getFiltersFromPreferences' or something in there, that would return a `QStringList` with those filters? Then here, you'd just iterate over the strings on that list, and append them. This way, we don't need to depend on the internals of that struct here

Reply:
Okay! I'll work on this

Reply:
Okay I added a new `QString getSelectedFiltersAsRegex()` function to InstanceCopyPrefs that looks like:

```C++
// Returns a single RegEx string of the selected folders/files to filter out (ex: ".minecraft/saves|.minecraft/mods")
QString InstanceCopyPrefs::getSelectedFiltersAsRegex() const
{
    QStringList filters;

    if(!copySaves)
        filters << "saves";

    if(!copyGameOptions)
        filters << "options.txt";

    if(!copyResourcePacks)
        filters << "resourcepacks" << "texturepacks";

    if(!copyShaderPacks)
        filters << "shaderpacks";

    if(!copyServers)
        filters << "servers.dat" << "servers.dat_old" << "server-resource-packs";

    if(!copyMods)
        filters << "coremods" << "mods" << "config";

    // If we have any filters to add, join them as a single regex string to return:
    if (!filters.isEmpty()) {
        const QString MC_ROOT = "[.]?minecraft/";
        // Ensure first filter starts with root, then join other filters with OR regex before root (ex: ".minecraft/saves|.minecraft/mods"):
        return MC_ROOT + filters.join("|" + MC_ROOT);
    }

    return "";
}
```

So that the constructor of InstanceCopyTask now looks like this:

```C++
InstanceCopyTask::InstanceCopyTask(InstancePtr origInstance, const InstanceCopyPrefs& prefs)
{
    m_origInstance = origInstance;
    m_keepPlaytime = prefs.keepPlaytime;

    QString filters = prefs.getSelectedFiltersAsRegex();
    if (!filters.isEmpty())
    {
        // Set regex filter:
        // FIXME: get this from the original instance type...
        auto matcherReal = new RegexpMatcher(filters);
        matcherReal->caseSensitive(false);
        m_matcher.reset(matcherReal);
    }
}
```

Is this good?

Reply:
sure, thx!